### ISSUE TITLE
Control plane not initialized in case of CAPD

### ISSUE THREAD

**Issue by AartiJivrajani** [kubernetes-sigs/cluster-api#5003]
**What steps did you take and what happened:**
I followed the 0.3 Quickstart guide here - https://release-0-3.cluster-api.sigs.k8s.io/user/quick-start.html for deploying a workload cluster using Docker
```
➜  capi-learn cat > kind-cluster-with-extramounts.yaml <<EOF
kind: Cluster
apiVersion: kind.x-k8s.io/v1alpha4
nodes:
- role: control-plane
  extraMounts:
    - hostPath: /var/run/docker.sock
      containerPath: /var/run/docker.sock
EOF
➜  capi-learn
➜  capi-learn kind create cluster --config kind-cluster-with-extramounts.yaml
Creating cluster "kind" ...
 ✓ Ensuring node image (kindest/node:v1.21.1) 🖼
⢄⡱ Preparing nodes 📦
 ✓ Preparing nodes 📦
 ✓ Writing configuration 📜
 ✓ Starting control-plane 🕹️
 ✓ Installing CNI 🔌
 ✓ Installing StorageClass 💾
Set kubectl context to "kind-kind"
You can now use your cluster with:

kubectl cluster-info --context kind-kind

Not sure what to do next? 😅  Check out https://kind.sigs.k8s.io/docs/user/quick-start/

➜  capi-learn clusterctl init --infrastructure docker
Fetching providers
Installing cert-manager Version="v1.1.0"
Waiting for cert-manager to be available...
Installing Provider="cluster-api" Version="v0.3.22" TargetNamespace="capi-system"
Installing Provider="bootstrap-kubeadm" Version="v0.3.22" TargetNamespace="capi-kubeadm-bootstrap-system"
Installing Provider="control-plane-kubeadm" Version="v0.3.22" TargetNamespace="capi-kubeadm-control-plane-system"
Installing Provider="infrastructure-docker" Version="v0.3.22" TargetNamespace="capd-system"

Your management cluster has been initialized successfully!

You can now create your first workload cluster by running the following:

  clusterctl config cluster [name] --kubernetes-version [version] | kubectl apply -f -


➜  capi-learn clusterctl config cluster capi-quickstart --flavor development \
  --kubernetes-version v1.21.1 \
  --control-plane-machine-count=3 \
  --worker-machine-count=3 \
  > capi-quickstart.yaml

➜  capi-learn kubectl apply -f capi-quickstart.yaml
cluster.cluster.x-k8s.io/capi-quickstart created
dockercluster.infrastructure.cluster.x-k8s.io/capi-quickstart created
kubeadmcontrolplane.controlplane.cluster.x-k8s.io/capi-quickstart-control-plane created
kubeadmconfigtemplate.bootstrap.cluster.x-k8s.io/capi-quickstart-md-0 created
machinedeployment.cluster.x-k8s.io/capi-quickstart-md-0 created
dockermachinetemplate.infrastructure.cluster.x-k8s.io/capi-quickstart-control-plane created
```
**What did you expect to happen:**
Expected the kubeadm controlplane to be initialized
```
➜  capi-learn kubectl get kubeadmcontrolplane --all-namespaces

NAMESPACE   NAME                            INITIALIZED   API SERVER AVAILABLE   VERSION   REPLICAS   READY   UPDATED   UNAVAILABLE
default     capi-quickstart-control-plane                                        v1.21.1   1                  1         1

# docker specific instructions
# Point the kubeconfig to the exposed port of the load balancer, rather than the inaccessible container IP.
sed -i -e "s/server:.*/server: https:\/\/$(docker port capi-quickstart-lb 6443/tcp | sed "s/0.0.0.0/127.0.0.1/")/g" ./capi-quickstart.kubeconfig

# Ignore the CA, because it is not signed for 127.0.0.1
sed -i -e "s/certificate-authority-data:.*/insecure-skip-tls-verify: true/g" ./capi-quickstart.kubeconfig

```

**Anything else you would like to add:**
```
➜  ~ docker ps -a
CONTAINER ID   IMAGE                                COMMAND                  CREATED              STATUS          PORTS                                  NAMES
106cdb2d66c7   kindest/node:v1.21.1                 "/usr/local/bin/entr…"   About a minute ago   Up 59 seconds   34649/tcp, 127.0.0.1:34649->6443/tcp   capi-quickstart-control-plane-fpj8r
ff02173c5a13   kindest/haproxy:v20210715-a6da3463   "haproxy -sf 7 -W -d…"   9 minutes ago        Up 9 minutes    41925/tcp, 0.0.0.0:41925->6443/tcp     capi-quickstart-lb
c7c29aa45558   kindest/node:v1.21.1                 "/usr/local/bin/entr…"   12 minutes ago       Up 12 minutes   127.0.0.1:63997->6443/tcp              kind-control-plane
```

```
➜  ~ kind get clusters
capi-quickstart
kind
```

```
➜  capi-learn kubectl --kubeconfig=./capi-quickstart.kubeconfig \
  apply -f https://docs.projectcalico.org/v3.15/manifests/calico.yaml --v=7
I0722 14:18:42.603404   17710 loader.go:372] Config loaded from file:  ./capi-quickstart.kubeconfig
I0722 14:18:42.604797   17710 round_trippers.go:432] GET https://127.0.0.1:41925/openapi/v2?timeout=32s
I0722 14:18:42.604815   17710 round_trippers.go:438] Request Headers:
I0722 14:18:42.604820   17710 round_trippers.go:442]     Accept: application/com.github.proto-openapi.spec.v2@v1.0+protobuf
I0722 14:18:42.604825   17710 round_trippers.go:442]     User-Agent: kubectl/v1.21.3 (darwin/amd64) kubernetes/ca643a4
I0722 14:18:42.606576   17710 round_trippers.go:457] Response Status:  in 1 milliseconds
I0722 14:18:42.606985   17710 request.go:959] Got a Retry-After 1s response for attempt 1 to https://127.0.0.1:41925/openapi/v2?timeout=32s
I0722 14:18:43.607469   17710 round_trippers.go:432] GET https://127.0.0.1:41925/openapi/v2?timeout=32s
I0722 14:18:43.607495   17710 round_trippers.go:438] Request Headers:
I0722 14:18:43.607504   17710 round_trippers.go:442]     Accept: application/com.github.proto-openapi.spec.v2@v1.0+protobuf
I0722 14:18:43.607512   17710 round_trippers.go:442]     User-Agent: kubectl/v1.21.3 (darwin/amd64) kubernetes/ca643a4
I0722 14:18:43.609909   17710 round_trippers.go:457] Response Status:  in 2 milliseconds
I0722 14:18:43.609991   17710 request.go:959] Got a Retry-After 1s response for attempt 2 to https://127.0.0.1:41925/openapi/v2?timeout=32s
I0722 14:18:44.613977   17710 round_trippers.go:432] GET https://127.0.0.1:41925/openapi/v2?timeout=32s
I0722 14:18:44.614004   17710 round_trippers.go:438] Request Headers:
I0722 14:18:44.614013   17710 round_trippers.go:442]     Accept: application/com.github.proto-openapi.spec.v2@v1.0+protobuf
I0722 14:18:44.614021   17710 round_trippers.go:442]     User-Agent: kubectl/v1.21.3 (darwin/amd64) kubernetes/ca643a4
I0722 14:18:44.616378   17710 round_trippers.go:457] Response Status:  in 2 milliseconds
I0722 14:18:44.616429   17710 request.go:959] Got a Retry-After 1s response for attempt 3 to https://127.0.0.1:41925/openapi/v2?timeout=32s
I0722 14:18:45.621343   17710 round_trippers.go:432] GET https://127.0.0.1:41925/openapi/v2?timeout=32s
I0722 14:18:45.621370   17710 round_trippers.go:438] Request Headers:
I0722 14:18:45.621379   17710 round_trippers.go:442]     Accept: application/com.github.proto-openapi.spec.v2@v1.0+protobuf
I0722 14:18:45.621387   17710 round_trippers.go:442]     User-Agent: kubectl/v1.21.3 (darwin/amd64) kubernetes/ca643a4
I0722 14:18:45.624151   17710 round_trippers.go:457] Response Status:  in 2 milliseconds
I0722 14:18:45.624203   17710 request.go:959] Got a Retry-After 1s response for attempt 4 to https://127.0.0.1:41925/openapi/v2?timeout=32s
I0722 14:18:46.628919   17710 round_trippers.go:432] GET https://127.0.0.1:41925/openapi/v2?timeout=32s
I0722 14:18:46.628961   17710 round_trippers.go:438] Request Headers:
I0722 14:18:46.628975   17710 round_trippers.go:442]     Accept: application/com.github.proto-openapi.spec.v2@v1.0+protobuf
I0722 14:18:46.628987   17710 round_trippers.go:442]     User-Agent: kubectl/v1.21.3 (darwin/amd64) kubernetes/ca643a4
I0722 14:18:46.632270   17710 round_trippers.go:457] Response Status:  in 3 milliseconds
I0722 14:18:46.632336   17710 request.go:959] Got a Retry-After 1s response for attempt 5 to https://127.0.0.1:41925/openapi/v2?timeout=32s
I0722 14:18:47.632483   17710 round_trippers.go:432] GET https://127.0.0.1:41925/openapi/v2?timeout=32s
I0722 14:18:47.632498   17710 round_trippers.go:438] Request Headers:
I0722 14:18:47.632502   17710 round_trippers.go:442]     User-Agent: kubectl/v1.21.3 (darwin/amd64) kubernetes/ca643a4
I0722 14:18:47.632506   17710 round_trippers.go:442]     Accept: application/com.github.proto-openapi.spec.v2@v1.0+protobuf
I0722 14:18:47.634522   17710 round_trippers.go:457] Response Status:  in 2 milliseconds
I0722 14:18:47.634560   17710 request.go:959] Got a Retry-After 1s response for attempt 6 to https://127.0.0.1:41925/openapi/v2?timeout=32s
I0722 14:18:48.635575   17710 round_trippers.go:432] GET https://127.0.0.1:41925/openapi/v2?timeout=32s
I0722 14:18:48.635601   17710 round_trippers.go:438] Request Headers:
I0722 14:18:48.635610   17710 round_trippers.go:442]     User-Agent: kubectl/v1.21.3 (darwin/amd64) kubernetes/ca643a4
I0722 14:18:48.635617   17710 round_trippers.go:442]     Accept: application/com.github.proto-openapi.spec.v2@v1.0+protobuf
I0722 14:18:48.638294   17710 round_trippers.go:457] Response Status:  in 2 milliseconds
I0722 14:18:48.638340   17710 request.go:959] Got a Retry-After 1s response for attempt 7 to https://127.0.0.1:41925/openapi/v2?timeout=32s
I0722 14:18:49.643436   17710 round_trippers.go:432] GET https://127.0.0.1:41925/openapi/v2?timeout=32s
I0722 14:18:49.643465   17710 round_trippers.go:438] Request Headers:
I0722 14:18:49.643474   17710 round_trippers.go:442]     Accept: application/com.github.proto-openapi.spec.v2@v1.0+protobuf
I0722 14:18:49.643482   17710 round_trippers.go:442]     User-Agent: kubectl/v1.21.3 (darwin/amd64) kubernetes/ca643a4
I0722 14:18:49.646352   17710 round_trippers.go:457] Response Status:  in 2 milliseconds
I0722 14:18:49.646400   17710 request.go:959] Got a Retry-After 1s response for attempt 8 to https://127.0.0.1:41925/openapi/v2?timeout=32s
I0722 14:18:50.650351   17710 round_trippers.go:432] GET https://127.0.0.1:41925/openapi/v2?timeout=32s
I0722 14:18:50.650371   17710 round_trippers.go:438] Request Headers:
I0722 14:18:50.650378   17710 round_trippers.go:442]     Accept: application/com.github.proto-openapi.spec.v2@v1.0+protobuf
I0722 14:18:50.650384   17710 round_trippers.go:442]     User-Agent: kubectl/v1.21.3 (darwin/amd64) kubernetes/ca643a4
I0722 14:18:50.652552   17710 round_trippers.go:457] Response Status:  in 2 milliseconds
I0722 14:18:50.652602   17710 request.go:959] Got a Retry-After 1s response for attempt 9 to https://127.0.0.1:41925/openapi/v2?timeout=32s
I0722 14:18:51.655566   17710 round_trippers.go:432] GET https://127.0.0.1:41925/openapi/v2?timeout=32s
I0722 14:18:51.655591   17710 round_trippers.go:438] Request Headers:
I0722 14:18:51.655600   17710 round_trippers.go:442]     Accept: application/com.github.proto-openapi.spec.v2@v1.0+protobuf
I0722 14:18:51.655608   17710 round_trippers.go:442]     User-Agent: kubectl/v1.21.3 (darwin/amd64) kubernetes/ca643a4
I0722 14:18:51.658528   17710 round_trippers.go:457] Response Status:  in 2 milliseconds
I0722 14:18:51.658590   17710 request.go:959] Got a Retry-After 1s response for attempt 10 to https://127.0.0.1:41925/openapi/v2?timeout=32s
I0722 14:18:52.661919   17710 round_trippers.go:432] GET https://127.0.0.1:41925/openapi/v2?timeout=32s
I0722 14:18:52.661938   17710 round_trippers.go:438] Request Headers:
I0722 14:18:52.661943   17710 round_trippers.go:442]     Accept: application/com.github.proto-openapi.spec.v2@v1.0+protobuf
I0722 14:18:52.661948   17710 round_trippers.go:442]     User-Agent: kubectl/v1.21.3 (darwin/amd64) kubernetes/ca643a4
I0722 14:18:52.663701   17710 round_trippers.go:457] Response Status:  in 1 milliseconds
I0722 14:18:52.665194   17710 helpers.go:216] server response object: [{
  "metadata": {},
  "status": "Failure",
  "message": "an error on the server (\"\") has prevented the request from succeeding",
  "reason": "InternalError",
  "details": {
    "causes": [
      {
        "reason": "UnexpectedServerResponse"
      }
    ],
    "retryAfterSeconds": 1
  },
  "code": 500
}]
F0722 14:18:52.666423   17710 helpers.go:115] Error from server (InternalError): an error on the server ("") has prevented the request from succeeding
goroutine 1 [running]:
k8s.io/kubernetes/vendor/k8s.io/klog/v2.stacks(0xc0001a4001, 0xc000722000, 0x97, 0xe8)
	/private/tmp/kubernetes-cli-20210716-2207-1d5c2f4/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/klog/v2/klog.go:1021 +0xb9
k8s.io/kubernetes/vendor/k8s.io/klog/v2.(*loggingT).output(0x6d8e620, 0xc000000003, 0x0, 0x0, 0xc000494070, 0x62e57cd, 0xa, 0x73, 0x4010800)
	/private/tmp/kubernetes-cli-20210716-2207-1d5c2f4/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/klog/v2/klog.go:970 +0x191
k8s.io/kubernetes/vendor/k8s.io/klog/v2.(*loggingT).printDepth(0x6d8e620, 0xc000000003, 0x0, 0x0, 0x0, 0x0, 0x2, 0xc0002a6af0, 0x1, 0x1)
	/private/tmp/kubernetes-cli-20210716-2207-1d5c2f4/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/klog/v2/klog.go:733 +0x16f
k8s.io/kubernetes/vendor/k8s.io/klog/v2.FatalDepth(...)
	/private/tmp/kubernetes-cli-20210716-2207-1d5c2f4/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/klog/v2/klog.go:1495
k8s.io/kubernetes/vendor/k8s.io/kubectl/pkg/cmd/util.fatal(0xc000d08000, 0x68, 0x1)
	/private/tmp/kubernetes-cli-20210716-2207-1d5c2f4/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/kubectl/pkg/cmd/util/helpers.go:93 +0x288
k8s.io/kubernetes/vendor/k8s.io/kubectl/pkg/cmd/util.checkErr(0x5d401a0, 0xc000566000, 0x5bbce88)
	/private/tmp/kubernetes-cli-20210716-2207-1d5c2f4/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/kubectl/pkg/cmd/util/helpers.go:188 +0x935
k8s.io/kubernetes/vendor/k8s.io/kubectl/pkg/cmd/util.CheckErr(...)
	/private/tmp/kubernetes-cli-20210716-2207-1d5c2f4/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/kubectl/pkg/cmd/util/helpers.go:115
k8s.io/kubernetes/vendor/k8s.io/kubectl/pkg/cmd/apply.NewCmdApply.func1(0xc0009c6840, 0xc000523c80, 0x0, 0x4)
	/private/tmp/kubernetes-cli-20210716-2207-1d5c2f4/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/kubectl/pkg/cmd/apply/apply.go:177 +0x73
k8s.io/kubernetes/vendor/github.com/spf13/cobra.(*Command).execute(0xc0009c6840, 0xc000523c40, 0x4, 0x4, 0xc0009c6840, 0xc000523c40)
	/private/tmp/kubernetes-cli-20210716-2207-1d5c2f4/_output/local/go/src/k8s.io/kubernetes/vendor/github.com/spf13/cobra/command.go:854 +0x2c2
k8s.io/kubernetes/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0xc000474dc0, 0xc0001a6180, 0xc0001a6120, 0x6)
	/private/tmp/kubernetes-cli-20210716-2207-1d5c2f4/_output/local/go/src/k8s.io/kubernetes/vendor/github.com/spf13/cobra/command.go:958 +0x375
k8s.io/kubernetes/vendor/github.com/spf13/cobra.(*Command).Execute(...)
	/private/tmp/kubernetes-cli-20210716-2207-1d5c2f4/_output/local/go/src/k8s.io/kubernetes/vendor/github.com/spf13/cobra/command.go:895
main.main()
	_output/local/go/src/k8s.io/kubernetes/cmd/kubectl/kubectl.go:49 +0x21d

goroutine 19 [chan receive]:
k8s.io/kubernetes/vendor/k8s.io/klog/v2.(*loggingT).flushDaemon(0x6d8e620)
	/private/tmp/kubernetes-cli-20210716-2207-1d5c2f4/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/klog/v2/klog.go:1164 +0x8b
created by k8s.io/kubernetes/vendor/k8s.io/klog/v2.init.0
	/private/tmp/kubernetes-cli-20210716-2207-1d5c2f4/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/klog/v2/klog.go:418 +0xdf

goroutine 10 [select]:
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0x5bbcda8, 0x5d404a0, 0xc0004b0000, 0x1, 0xc0001b0ae0)
	/private/tmp/kubernetes-cli-20210716-2207-1d5c2f4/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:167 +0x118
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil(0x5bbcda8, 0x12a05f200, 0x0, 0x1, 0xc0001b0ae0)
	/private/tmp/kubernetes-cli-20210716-2207-1d5c2f4/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0x98
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.Until(0x5bbcda8, 0x12a05f200, 0xc0001b0ae0)
	/private/tmp/kubernetes-cli-20210716-2207-1d5c2f4/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:90 +0x4d
created by k8s.io/kubernetes/vendor/k8s.io/kubectl/pkg/util/logs.InitLogs
	/private/tmp/kubernetes-cli-20210716-2207-1d5c2f4/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/kubectl/pkg/util/logs/logs.go:51 +0x96
```

logs:
[capd-controller-manager-5579d75d6f-s4wpc](https://codeshare.io/Rb7zW4)
[capi-controller-manager-7d56fdcb8f-2lkjr](https://codeshare.io/X8LKBj)
[capi-kubeadm-bootstrap-controller-manager-98c5486db-smb6b](https://codeshare.io/8pXrgJ)
[capi-kubeadm-control-plane-controller-manager-6c8fdc9565-5j49c](https://codeshare.io/zylg67)
[Webhook-system: capi-controller-manager-798ccdd5b8-sj6js](https://codeshare.io/3AyoxB)

**Environment:**

- Cluster-api version: 0.3.22
- Minikube/KIND version: kind 0.11
- Kubernetes version: (use `kubectl version`): 1.21.3 
- OS (e.g. from `/etc/os-release`): macOS 11.3.1

**Comment by sbueringer** [comment]
Thank you for the very detailed issue :). I'll take a look

**Comment by sbueringer** [comment]
We had the same issue on our main branch. 

> Starting with kind v0.11, kind images will have cgroup-root=/kubelet hard-coded to fix an issue with nested cgroup hierarchies. The hierarchy below /kubelet is not compatible with systemd cgroupfs (in contrast to /).
Starting with kubeadm v1.21, kubeadm will default to cgroupDriver systemd.
> Thus, the combination of images from kind >=v0.11 and kubeadm >=v1.21 are not compatible. To solve this problem we will pin cgroupDriver to cgroupfs until kind supports the systemd hierarchy with the /kubelet cgroup-root.
> https://github.com/kubernetes-sigs/cluster-api/pull/4614

tl;dr To get it working (for now) you have to add `cgroup-driver: cgroupfs` to all `kubeletExtraArgs` in `capi-quickstart.yaml` manually.


@fabriziopandini Should we cherry-pick the fix to the release-0.3 branch? I think we can be reasonable sure that it works for older version because of our test coverage on main starting from v1.18. I think we should do it, otherwise the CAPD v0.3.x quickstart will fail for everyone using Kubernetes >= v1.21.0 which is bad :/

**Comment by fabriziopandini** [comment]
+1 to backport the fix (and eventually fix released manifests) given that I expect more people using 1.21 or 1.22 soon

**Comment by vincepri** [comment]
@AartiJivrajani #5009 backported the fix, please let us know if the problem persists and feel free to reopen if necessary.

/close
/milestone v0.3
